app/src/main/java/com/github/jorgecastillo/kotlinandroid/data/datasource/remote/MarvelNetworkDataSource.kt

```diff
@@ -24,35 +25,31 @@ import java.net.HttpURLConnection
  * word in the title. Yep, I wanted to retrieve Avengers but the Marvel API is a bit weird
  * sometimes.
  */
+fun exceptionAsCharacterError(e: Throwable): CharacterError =
+        when (e) {
+            is MarvelAuthApiException -> AuthenticationError
+            is MarvelApiException ->
+                if (e.httpCode == HttpURLConnection.HTTP_NOT_FOUND) NotFoundError
+                else UnknownServerError(Option.Some(e))
+            else -> UnknownServerError((Option.Some(e)))
+        }
 
-fun fetchAllHeroes() = Reader.ask<GetHeroesContext>().map { ctx ->
-  Future {
-    try {
-      val query = CharactersQuery.Builder.create().withOffset(0).withLimit(50).build()
-      Right<List<CharacterDto>>(ctx.apiClient.getAll(query).response.characters)
-    } catch (e: MarvelAuthApiException) {
-      Left(AuthenticationError())
-    } catch (e: MarvelApiException) {
-      if (e.httpCode == HttpURLConnection.HTTP_NOT_FOUND) {
-        Left(NotFoundError())
-      } else {
-        Left(UnknownServerError())
-      }
-    }
-  }
-}
 
-fun fetchHeroesFromAvengerComics() = fetchAllHeroes().map { future ->
-  future.map {
-    when (it) {
-      is Right -> it.map {
-        it.filter {
-          it.comics.items.map { it.name }.filter {
-            it.contains("Avenger", true)
-          }.count() > 0
+inline fun <reified F> fetchAllHeroes(C : Control<F> = control()): HK<F, List<CharacterDto>> =
+        C.binding {
+            val query = CharactersQuery.Builder.create().withOffset(0).withLimit(50).build()
+            val ctx = !C.ask()
```

Comment:
Could agree, but that does not mean we should probably try to give them the knowledge in an easier way so the learning curve is lower. Why not just using "bind" here, to start with? Minor change, but probably helpful.